p2p/querier.go

```diff
@@ -2,61 +2,258 @@ package p2p
+					"majThreshHold",
+					majThreshHold,
+					"currThreshold",
+					currThreshold,
+				)
+				return confirms, nil
+			}
+			q.logger.Debug(
+				"found DataCommitmentConfirms",
+				"total_power",
+				currThreshold,
+				"number_of_confirms",
+				len(confirms),
+				"missing_confirms",
+				len(previousValset.Members)-len(confirms),
+			)
+		}
+		// TODO: make the sleep configurable
+		// TODO make it as a parameter and use ticker
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// QueryTwoThirdsValsetConfirms queries two thirds or more of valset confirms from the
+// P2P network. The method will not return unless it finds more than two thirds, or it times out.
+// No validation is required to be done at this level because the P2P validators defined at
+// `p2p/validators.go` will make sure that the values queried from the DHT are valid.
+// Should return an empty slice and no error if it couldn't find enough signatures.
+// For valsets, generally the first valset, whose nonce is 1, is never signed by the network.
+// Thus, the call to this method for the genesis valset will only timeout.
+func (q Querier) QueryTwoThirdsValsetConfirms(
+	ctx context.Context,
+	timeout time.Duration,
+	valset celestiatypes.Valset,
+) ([]types.ValsetConfirm, error) {
+	// create a map to easily search for power
+	vals := make(map[string]celestiatypes.BridgeValidator)
+	for _, val := range valset.Members {
+		vals[val.GetEvmAddress()] = val
+	}
+
+	majThreshHold := valset.TwoThirdsThreshold()
+	t := time.After(timeout)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil //nolint:nilnil
+		// TODO: remove this extra case, and we can instead rely on the caller to pass a context with a timeout
+		case <-t:
+			return nil, pkgerrors.Wrap(
+				ErrNotEnoughValsetConfirms,
+				fmt.Sprintf("failure to query for majority validator set confirms: timout %s", timeout),
+			)
+		default:
+			confirms, err := q.QueryValsetConfirms(ctx, valset)
+			if err != nil {
+				return nil, err
+			}
+
+			currThreshold := uint64(0)
+			for _, valsetConfirm := range confirms {
+				val, has := vals[valsetConfirm.EthAddress]
+				if !has {
+					q.logger.Debug(
+						fmt.Sprintf(
+							"valSetConfirm signer not found in stored validator set: address %s nonce %d",
+							val.EvmAddress,
+							valset.Nonce,
+						))
+					continue
+				}
+				currThreshold += val.Power
+			}
+
+			if currThreshold >= majThreshHold {
+				q.logger.Debug("found enough valset confirms to be relayed",
+					"majThreshHold",
+					majThreshHold,
+					"currThreshold",
+					currThreshold,
+				)
+				return confirms, nil
+			}
+			q.logger.Debug(
+				"found ValsetConfirms",
+				"nonce",
+				valset.Nonce,
+				"total_power",
+				currThreshold,
+				"number_of_confirms",
+				len(confirms),
+				"missing_confirms",
+				len(valset.Members)-len(confirms),
+			)
+		}
+		// TODO: make the timeout configurable
+		time.Sleep(10 * time.Second)
```

Comment:
what is this for?

Reply:
Just to keep a delay between the different calls to querying 2/3. I wanna change it with a ticker, what do you think? every 10sec

Reply:
I see, a ticker could be more explicit yeah

Reply:
https://github.com/celestiaorg/orchestrator-relayer/issues/119